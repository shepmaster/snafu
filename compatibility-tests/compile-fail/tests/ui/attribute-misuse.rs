mod enum_misuse {
    use snafu::prelude::*;

    #[derive(Debug, Snafu)]
    #[snafu(display("display should not work here"))]
    #[snafu(source(from(XXXX, Box::new)))]
    #[snafu(source(true))]
    #[snafu(backtrace)]
    #[snafu(context)]
    #[snafu(context(false))]
    #[snafu(context(name(Name)))]
    #[snafu(implicit)]
    #[snafu(provide)]
    #[snafu(provide(u8 => 0))]
    #[snafu(transparent)]
    enum EnumError {
        AVariant,
    }
}

mod variant_misuse {
    use snafu::prelude::*;

    #[derive(Debug, Snafu)]
    enum EnumError {
        // Make sure we catch the error in the second attribute
        #[snafu(display("an error variant"), source(from(XXXX, Box::new)))]
        #[snafu(source)]
        #[snafu(backtrace)]
        #[snafu(crate_root(XXXX))]
        #[snafu(implicit)]
        #[snafu(provide)]
        AVariant,
    }
}

mod field_misuse {
    use snafu::prelude::*;

    #[derive(Debug, Snafu)]
    enum EnumError {
        AVariant {
            #[snafu(display("display should not work here"))]
            #[snafu(visibility(pub))]
            #[snafu(source(false))]
            #[snafu(source(from(XXXX, Box::new)))]
            #[snafu(context)]
            #[snafu(context(false))]
            #[snafu(context(suffix(Suffix)))]
            #[snafu(context(name(Name)))]
            #[snafu(crate_root(XXXX))]
            #[snafu(transparent)]
            source: String,

            #[snafu(provide(false))]
            #[snafu(provide(u8 => 0))]
            not_a_source: bool,
        },
    }
}

mod struct_misuse {
    use snafu::prelude::*;

    #[derive(Debug, Snafu)]
    enum UsableError {}

    #[derive(Debug, Snafu)]
    #[snafu(display("display should not work here"))]
    #[snafu(source(from(UsableError, Box::new)))]
    #[snafu(visibility(pub))]
    #[snafu(source(true))]
    #[snafu(backtrace)]
    #[snafu(context)]
    #[snafu(context(false))]
    #[snafu(context(suffix(Suffix)))]
    #[snafu(context(name(Name)))]
    #[snafu(implicit)]
    #[snafu(provide)]
    #[snafu(provide(u8 => 0))]
    #[snafu(transparent)]
    struct StructError(Box<UsableError>);

    mod field_misuse {
        use snafu::prelude::*;

        #[derive(Debug, Snafu)]
        struct Opaque(
            #[snafu(backtrace)]
            #[snafu(context)]
            #[snafu(context(false))]
            #[snafu(context(suffix(Suffix)))]
            #[snafu(context(name(Name)))]
            #[snafu(crate_root(nowhere))]
            #[snafu(display("display should not work here"))]
            #[snafu(implicit)]
            #[snafu(module)]
            #[snafu(provide)]
            #[snafu(source)]
            #[snafu(transparent)]
            #[snafu(visibility(pub))]
            #[snafu(whatever)]
            Box<InnerError>,
        );

        #[derive(Debug, Snafu)]
        struct InnerError;
    }
}

fn main() {}
