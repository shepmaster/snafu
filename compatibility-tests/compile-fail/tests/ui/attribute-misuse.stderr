error: `display` attribute is only valid on enum variants or structs with named fields, not on an enum
 --> tests/ui/attribute-misuse.rs:5:13
  |
5 |     #[snafu(display("display should not work here"))]
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `source(from)` attribute is only valid on enum variant or struct fields with a name, not on an enum
 --> tests/ui/attribute-misuse.rs:6:13
  |
6 |     #[snafu(source(from(XXXX, Box::new)))]
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `source(bool)` attribute is only valid on enum variant or struct fields with a name, not on an enum
 --> tests/ui/attribute-misuse.rs:7:13
  |
7 |     #[snafu(source(true))]
  |             ^^^^^^^^^^^^

error: `backtrace` attribute is only valid on enum variant or struct fields with a name, not on an enum
 --> tests/ui/attribute-misuse.rs:8:13
  |
8 |     #[snafu(backtrace)]
  |             ^^^^^^^^^

error: `context(bool)` attribute is only valid on enum variants or structs with named fields, not on an enum
 --> tests/ui/attribute-misuse.rs:9:13
  |
9 |     #[snafu(context)]
  |             ^^^^^^^

error: `context(bool)` attribute is only valid on enum variants or structs with named fields, not on an enum
  --> tests/ui/attribute-misuse.rs:10:13
   |
10 |     #[snafu(context(false))]
   |             ^^^^^^^^^^^^^^

error: `implicit` attribute is only valid on enum variant or struct fields with a name, not on an enum
  --> tests/ui/attribute-misuse.rs:11:13
   |
11 |     #[snafu(implicit)]
   |             ^^^^^^^^

error: `provide` attribute is only valid on enum variant or struct fields with a name, not on an enum
  --> tests/ui/attribute-misuse.rs:12:13
   |
12 |     #[snafu(provide)]
   |             ^^^^^^^

error: `provide(type => expression)` attribute is only valid on enum variants, structs with named fields, or tuple structs, not on an enum
  --> tests/ui/attribute-misuse.rs:13:13
   |
13 |     #[snafu(provide(u8 => 0))]
   |             ^^^^^^^^^^^^^^^^

error: `transparent` attribute is only valid on enum variants or structs with named fields, not on an enum
  --> tests/ui/attribute-misuse.rs:14:13
   |
14 |     #[snafu(transparent)]
   |             ^^^^^^^^^^^

error: `source` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> tests/ui/attribute-misuse.rs:26:46
   |
26 |         #[snafu(display("an error variant"), source(from(XXXX, Box::new)))]
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `source` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> tests/ui/attribute-misuse.rs:27:17
   |
27 |         #[snafu(source)]
   |                 ^^^^^^

error: `backtrace` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> tests/ui/attribute-misuse.rs:28:17
   |
28 |         #[snafu(backtrace)]
   |                 ^^^^^^^^^

error: `crate_root` attribute is only valid on an enum or a struct, not on an enum variant
  --> tests/ui/attribute-misuse.rs:29:17
   |
29 |         #[snafu(crate_root(XXXX))]
   |                 ^^^^^^^^^^^^^^^^

error: `implicit` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> tests/ui/attribute-misuse.rs:30:17
   |
30 |         #[snafu(implicit)]
   |                 ^^^^^^^^

error: `provide` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> tests/ui/attribute-misuse.rs:31:17
   |
31 |         #[snafu(provide)]
   |                 ^^^^^^^

error: `display` attribute is only valid on enum variants or structs with named fields, not on a field
  --> tests/ui/attribute-misuse.rs:42:21
   |
42 |             #[snafu(display("display should not work here"))]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `visibility` attribute is only valid on an enum, enum variants, or a struct with named fields, not on a field
  --> tests/ui/attribute-misuse.rs:43:21
   |
43 |             #[snafu(visibility(pub))]
   |                     ^^^^^^^^^^^^^^^

error: Incompatible attributes [`source(false)`, `source(from)`] specified on a field
  --> tests/ui/attribute-misuse.rs:45:21
   |
45 |             #[snafu(source(from(XXXX, Box::new)))]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a field
  --> tests/ui/attribute-misuse.rs:46:21
   |
46 |             #[snafu(context)]
   |                     ^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a field
  --> tests/ui/attribute-misuse.rs:47:21
   |
47 |             #[snafu(context(false))]
   |                     ^^^^^^^^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a field
  --> tests/ui/attribute-misuse.rs:48:21
   |
48 |             #[snafu(context(suffix(Suffix)))]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^

error: `crate_root` attribute is only valid on an enum or a struct, not on a field
  --> tests/ui/attribute-misuse.rs:49:21
   |
49 |             #[snafu(crate_root(XXXX))]
   |                     ^^^^^^^^^^^^^^^^

error: `transparent` attribute is only valid on enum variants or structs with named fields, not on a field
  --> tests/ui/attribute-misuse.rs:50:21
   |
50 |             #[snafu(transparent)]
   |                     ^^^^^^^^^^^

error: `provide(false)` attribute is only valid on a field named "source" or "backtrace", not on other fields
  --> tests/ui/attribute-misuse.rs:53:21
   |
53 |             #[snafu(provide(false))]
   |                     ^^^^^^^^^^^^^^

error: `provide(type => expression)` attribute is only valid on enum variants, structs with named fields, or tuple structs, not on a field
  --> tests/ui/attribute-misuse.rs:54:21
   |
54 |             #[snafu(provide(u8 => 0))]
   |                     ^^^^^^^^^^^^^^^^

error: `display` attribute is only valid on enum variants or structs with named fields, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:67:13
   |
67 |     #[snafu(display("display should not work here"))]
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `visibility` attribute is only valid on an enum, enum variants, or a struct with named fields, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:69:13
   |
69 |     #[snafu(visibility(pub))]
   |             ^^^^^^^^^^^^^^^

error: `source(bool)` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:70:13
   |
70 |     #[snafu(source(true))]
   |             ^^^^^^^^^^^^

error: `backtrace` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:71:13
   |
71 |     #[snafu(backtrace)]
   |             ^^^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:72:13
   |
72 |     #[snafu(context)]
   |             ^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:73:13
   |
73 |     #[snafu(context(false))]
   |             ^^^^^^^^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:74:13
   |
74 |     #[snafu(context(suffix(Suffix)))]
   |             ^^^^^^^^^^^^^^^^^^^^^^^

error: `implicit` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:75:13
   |
75 |     #[snafu(implicit)]
   |             ^^^^^^^^

error: `provide` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:76:13
   |
76 |     #[snafu(provide)]
   |             ^^^^^^^

error: `transparent` attribute is only valid on enum variants or structs with named fields, not on a tuple struct
  --> tests/ui/attribute-misuse.rs:78:13
   |
78 |     #[snafu(transparent)]
   |             ^^^^^^^^^^^

error: `backtrace` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:86:21
   |
86 |             #[snafu(backtrace)]
   |                     ^^^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:87:21
   |
87 |             #[snafu(context)]
   |                     ^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:88:21
   |
88 |             #[snafu(context(false))]
   |                     ^^^^^^^^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:89:21
   |
89 |             #[snafu(context(suffix(Suffix)))]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^

error: `crate_root` attribute is only valid on an enum or a struct, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:90:21
   |
90 |             #[snafu(crate_root(nowhere))]
   |                     ^^^^^^^^^^^^^^^^^^^

error: `display` attribute is only valid on enum variants or structs with named fields, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:91:21
   |
91 |             #[snafu(display("display should not work here"))]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `implicit` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:92:21
   |
92 |             #[snafu(implicit)]
   |                     ^^^^^^^^

error: `module` attribute is only valid on an enum or structs with named fields, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:93:21
   |
93 |             #[snafu(module)]
   |                     ^^^^^^

error: `provide` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:94:21
   |
94 |             #[snafu(provide)]
   |                     ^^^^^^^

error: `source` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:95:21
   |
95 |             #[snafu(source)]
   |                     ^^^^^^

error: `transparent` attribute is only valid on enum variants or structs with named fields, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:96:21
   |
96 |             #[snafu(transparent)]
   |                     ^^^^^^^^^^^

error: `visibility` attribute is only valid on an enum, enum variants, or a struct with named fields, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:97:21
   |
97 |             #[snafu(visibility(pub))]
   |                     ^^^^^^^^^^^^^^^

error: `whatever` attribute is only valid on enum variants or structs with named fields, not on a tuple struct field
  --> tests/ui/attribute-misuse.rs:98:21
   |
98 |             #[snafu(whatever)]
   |                     ^^^^^^^^
